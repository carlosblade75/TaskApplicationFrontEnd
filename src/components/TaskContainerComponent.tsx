import React, {useState, useEffect} from 'react';
import { TaskService } from '../services/TaskService'
import { ITask } from '../Interface/ITask'
import { TaskListComponent } from './TaskListComponent';
import { TaskContext } from '../Context/TaskContext'

export const TaskContainerComponent = () => {

  const defaultProps = {
    pendingTaskList : [],
    compelteTaskList : [],
  };

  const [taskList, setTaskList] = useState(defaultProps);
  const [timeToRefresh, setTimeToRefresh] = useState(Date.now());
  const [descriptionTask, setDescriptionTask] = useState('')

  const fetchDataAsync = async () => {
    return  await TaskService.GetAllTask();
  } 

  const handleCheckUpdate = async (task:ITask) => {

    const {success, messageError} = await TaskService.UpdateTask(task);

    if (success) {
      setTimeToRefresh(Date.now());
    }
    else {
      // Just show the message Error in the consolo at the moment
      console.log(messageError);
    }
   
  }

  const handleAddButton = async () => {

    const task:ITask = {isCompleted :false, id : 0, description : descriptionTask};

    await TaskService.AddTask(task);

    setDescriptionTask('');
    
    setTimeToRefresh(Date.now());
  }

  const handleChangeDescription = (e: React.FormEvent<HTMLInputElement>) => {

    setDescriptionTask(e.currentTarget.value)
  }

  useEffect(() => {
    
    fetchDataAsync().then( data => {
      
      const pendingTask = data.filter( (task:ITask) => !task.isCompleted);

      const completedTask = data.filter( (task:ITask) => task.isCompleted);
       
      setTaskList({pendingTaskList: pendingTask, compelteTaskList: completedTask});

    });

  }, [timeToRefresh])

  return (

      <div>

        <TaskContext.Provider value= {{ checkUpdate : handleCheckUpdate }} >

          <TaskListComponent title="Pending Tasks" taskList={taskList.pendingTaskList} />

          <TaskListComponent title="Completed Tasks" taskList={taskList.compelteTaskList} />
          
          <input  type="text" 
                  value={descriptionTask} 
                  onChange = { handleChangeDescription }
                  placeholder="Add new task"/> <button onClick= { handleAddButton }> Add Task</button>

        </TaskContext.Provider>

      </div>
  )
}

