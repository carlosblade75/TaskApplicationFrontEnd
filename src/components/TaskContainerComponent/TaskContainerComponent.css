.containerTask {
    margin: 20px;
}

.containerAdd {
    margin-top: 50px;
}

.inputTask {
    width: 650px;
    padding: 5px;
}

.btnAddTask {
    width: 100px;
    margin-left: 10px;
}
