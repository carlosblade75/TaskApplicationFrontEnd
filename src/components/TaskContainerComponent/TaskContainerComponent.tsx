import React, {useState, useEffect} from 'react';
import { TaskService } from '../../services/TaskService'
import { ITask } from '../../Interface/ITask'
import { TaskListComponent } from '../../components/TaskListComponent/TaskListComponent'
import { TaskContext } from '../../Context/TaskContext'
import  './TaskContainerComponent.css'

export const TaskContainerComponent = () => {

  const defaultProps = {
    pendingTaskList : [],
    compelteTaskList : [],
  };

  const [taskList, setTaskList] = useState(defaultProps);
  const [timeToRefresh, setTimeToRefresh] = useState(Date.now());
  const [descriptionTask, setDescriptionTask] = useState('')

  const fetchDataAsync = async () => {

    try {
      return  await TaskService.GetAllTask();
    }
    catch(error) {
      console.error(error);
    }

  } 

  const handleCheckUpdate = async (task:ITask) => {

    try {
      
      const {success, messageError} = await TaskService.UpdateTask(task);

      if (success) {
        setTimeToRefresh(Date.now());
      }
      else {
        console.error(messageError);
      }

    }
    catch(error) {
      console.error(error);
    }
   
  }

  const  AddTaskService = async(task) =>  {

    await TaskService.AddTask(task);
  }

  const handleAddButton = async () => {

    try {

      if (descriptionTask.length > 0) {

        const task:ITask = {isCompleted :false, id : 0, description : descriptionTask};
        
        AddTaskService(task)

        setDescriptionTask('');
        
        setTimeToRefresh(Date.now());
      }
    }
    catch(error) {
      console.error(error);
    }

  }

  const handleChangeDescription = (e: React.FormEvent<HTMLInputElement>) => {
    setDescriptionTask(e.currentTarget.value)
  }

  useEffect(() => {
    
    fetchDataAsync().then( data => {
      
      const pendingTask = data.filter( (task:ITask) => !task.isCompleted);

      const completedTask = data.filter( (task:ITask) => task.isCompleted);
       
      setTaskList({pendingTaskList: pendingTask, completedTaskList: completedTask});

    });

  }, [timeToRefresh])

  return (

      <div className="container-fluid containerTask">

        <div className="row">

          <TaskContext.Provider value= {{ checkUpdate : handleCheckUpdate }} >

          <div className="col-6">
            <TaskListComponent title="Pending Tasks" taskList={taskList.pendingTaskList} />
          </div>

          <div className="col-6">
            <TaskListComponent title="Completed Tasks" taskList={taskList.completedTaskList} />
          </div>

          </TaskContext.Provider>

        </div>

        <hr></hr>

        <div className="row containerAdd">
          
          <input  type="text" 
                    value={descriptionTask} 
                    onChange = { handleChangeDescription }
                    placeholder="Add new task"
                    className="form-control inputTask"
          /> 

          <button className="btn btn-success btnAddTask" 
                  onClick= { handleAddButton }> 
                  Add Task
          </button>

        </div>

      </div>
  )
}

