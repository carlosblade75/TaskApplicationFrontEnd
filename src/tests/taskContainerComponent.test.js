import React from 'react'
import '@testing-library/jest-dom'
import { TaskContainerComponent } from "../components/TaskContainerComponent/TaskContainerComponent"
import { shallow } from 'enzyme'

describe('Testing for <TaskContainerComponent/>', () => {
    
    let wrapper = shallow( <TaskContainerComponent /> )

    beforeEach( () => {

        jest.clearAllMocks()
        wrapper = shallow( <TaskContainerComponent /> )
    })

    test('should show <TaskContainerComponent/> properly', () => {
        expect(  wrapper ).toMatchSnapshot()
    })

    test('should show two TaskListComponents', () => {
        expect (wrapper.find('TaskListComponent').length ).toBe( 2 )
    })

    test('should change the value of the input', () => {
        
        const setDescriptionTask = jest.fn()

        const value= 'Task new'

        const e = { currentTarget : { value: 'Task new' } }

        wrapper.find('input').simulate('change', e)

        expect( wrapper.find('input').prop('value') ).toBe( value )
    })

    test('should not update the description task method', () => {

        const setDescriptionTask = jest.fn()
        const setTimeToRefresh = jest.fn()

        wrapper.find('button').simulate('click')

        expect( wrapper.find('input').prop('value') ).toBe( '' )
        expect( setDescriptionTask ).not.toHaveBeenCalled()
        expect( setTimeToRefresh ).not.toHaveBeenCalled()
    })

    test('should empty the input once the add has been added', () => {

        const handleAddButton = jest.fn()

        const value= 'Task new'

        const e = { currentTarget : { value } }

        wrapper.find('input').simulate('change', e)

        wrapper.find('button').simulate('click')

        expect( wrapper.find('input').prop('value') ).toBe( '' )
    })
  
})
