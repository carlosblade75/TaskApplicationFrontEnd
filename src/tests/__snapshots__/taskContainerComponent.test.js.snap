// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testing for <TaskContainerComponent/> should show <TaskContainerComponent/> properly 1`] = `
<div
  className="container-fluid containerTask"
>
  <div
    className="row"
  >
    <ContextProvider
      value={
        Object {
          "checkUpdate": [Function],
        }
      }
    >
      <div
        className="col-6"
      >
        <TaskListComponent
          taskList={Array []}
          title="Pending Tasks"
        />
      </div>
      <div
        className="col-6"
      >
        <TaskListComponent
          taskList={Array []}
          title="Completed Tasks"
        />
      </div>
    </ContextProvider>
  </div>
  <hr />
  <div
    className="row containerAdd"
  >
    <input
      className="form-control inputTask"
      onChange={[Function]}
      placeholder="Add new task"
      type="text"
      value=""
    />
    <button
      className="btn btn-success btnAddTask"
      onClick={[Function]}
    >
      Add Task
    </button>
  </div>
</div>
`;
